contracts/FindToken.cdc

```diff
@@ -0,0 +1,188 @@
+
+        pub var balance: UFix64 
+        
+        init(balance: UFix64) {
+            self.balance=balance
+        }
+
+        destroy(){
+            FindToken.totalSupply = FindToken.totalSupply - self.balance
+            emit TokensBurned(amount: self.balance)
+        }
+
+        pub fun withdraw(amount: UFix64) : @FungibleToken.Vault {
+            self.balance = self.balance - amount 
+            emit TokensWithdrawn(amount: amount, from: self.owner?.address)
+            return <-create Vault(balance: amount)
+        }
+
+        pub fun deposit(from: @FungibleToken.Vault) {
+            let vault <- from as! @Vault 
+            self.balance = self.balance + vault.balance
+            emit TokensDeposited(amount: vault.balance, to: self.owner?.address)
+            vault.balance = 0.0 
+            destroy vault
+        }
+
+        pub fun getViews() : [Type] {
+            return [
+                Type<FindRewardToken.FTVaultData>() , 
+                Type<&FindToken.Vault{FungibleToken.Receiver}>() ,
+                Type<&FindToken.Vault{FungibleToken.Balance}>() , 
+                Type<&FindToken.Vault{FindRewardToken.VaultViews}>() 
+            ]
+        }
+
+        pub fun resolveView(_ view: Type) : AnyStruct? {
+            switch view {
+                case Type<FindRewardToken.FTVaultData>() :
+                    return FindRewardToken.FTVaultData(
+                            tokenAlias: FindToken.tokenAlias,
+                            storagePath: FindToken.vaultStoragePath,
+                            receiverPath: FindToken.receiverPublicPath,
+                            balancePath: FindToken.balancePublicPath,
+                            providerPath: FindToken.providerPath,
+                            findRewardPath: FindToken.findRewardPath,
+                            vaultType: FindToken.getTokenType(),
+                            receiverType: Type<&FindToken.Vault{FungibleToken.Receiver, FindRewardToken.VaultViews}>(),
+                            balanceType: Type<&FindToken.Vault{FungibleToken.Balance, FindRewardToken.VaultViews}>(),
+                            providerType: Type<&FindToken.Vault{FungibleToken.Provider, FindRewardToken.VaultViews}>(),
+                            findRewardType: Type<&FindToken.Vault{FindRewardToken.FindReward, FindRewardToken.VaultViews}>(),
+                            createEmptyVault: FindToken.createEmptyVaultFN()
+                        )
+                
+                case Type<&FindToken.Vault{FungibleToken.Receiver}>() :
+                    return (&self as &{FungibleToken.Receiver}?)!
+
+                case Type<&FindToken.Vault{FungibleToken.Balance}>() :
+                    return (&self as &{FungibleToken.Balance}?)!
+
+                default : 
+                    return nil 
+            }
+        }
+
+        pub fun reward(name: String, receiver: Address, task: String) : FindRewardToken.rewardPaid {
```

Comment:
This method should not pay out or emit events, it should return rewardPaid that has type and amount, rename to Reward. And when we call reward we will do the actual rewarding from a central place.